File: backend/src/main/java/com/condocompare/parceiros/service/ParceiroService.java
package com.condocompare.parceiros.service;

import com.condocompare.parceiros.dto.*;
import com.condocompare.parceiros.entity.CategoriaParceiro;
import com.condocompare.parceiros.entity.Parceiro;
import com.condocompare.parceiros.repository.ParceiroRepository;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

@Service
@Transactional(readOnly = true)
public class ParceiroService {

    private final ParceiroRepository parceiroRepository;

    public ParceiroService(ParceiroRepository parceiroRepository) {
        this.parceiroRepository = parceiroRepository;
    }

    @Transactional
    public ParceiroResponse create(CreateParceiroRequest request) {
        Parceiro parceiro = new Parceiro();
        mapRequestToEntity(request, parceiro);
        parceiro.setAtivo(true);
        parceiro.setVerificado(false);
        parceiro.setAvaliacao(null);
        parceiro.setTotalAvaliacoes(0);

        Parceiro saved = parceiroRepository.save(parceiro);
        return ParceiroResponse.from(saved);
    }

    @Transactional
    public ParceiroResponse update(UUID id, UpdateParceiroRequest request) {
        Parceiro parceiro = parceiroRepository.findById(id)
            .orElseThrow(() -> new EntityNotFoundException("Parceiro nao encontrado: " + id));

        if (request.nome() != null) parceiro.setNome(request.nome());
        if (request.nomeFantasia() != null) parceiro.setNomeFantasia(request.nomeFantasia());
        if (request.cnpj() != null) parceiro.setCnpj(request.cnpj());
        if (request.cpf() != null) parceiro.setCpf(request.cpf());
        if (request.email() != null) parceiro.setEmail(request.email());
        if (request.telefone() != null) parceiro.setTelefone(request.telefone());
        if (request.celular() != null) parceiro.setCelular(request.celular());
        if (request.website() != null) parceiro.setWebsite(request.website());
        if (request.endereco() != null) parceiro.setEndereco(request.endereco());
        if (request.numero() != null) parceiro.setNumero(request.numero());
        if (request.complemento() != null) parceiro.setComplemento(request.complemento());
        if (request.bairro() != null) parceiro.setBairro(request.bairro());
        if (request.cidade() != null) parceiro.setCidade(request.cidade());
        if (request.estado() != null) parceiro.setEstado(request.estado());
        if (request.cep() != null) parceiro.setCep(request.cep());
        if (request.categorias() != null && !request.categorias().isEmpty()) {
            parceiro.setCategorias(request.categorias());
        }
        if (request.descricaoServicos() != null) parceiro.setDescricaoServicos(request.descricaoServicos());
        if (request.areaAtuacao() != null) parceiro.setAreaAtuacao(request.areaAtuacao());
        if (request.contatoNome() != null) parceiro.setContatoNome(request.contatoNome());
        if (request.contatoCargo() != null) parceiro.setContatoCargo(request.contatoCargo());
        if (request.observacoes() != null) parceiro.setObservacoes(request.observacoes());
        if (request.logoUrl() != null) parceiro.setLogoUrl(request.logoUrl());
        if (request.ativo() != null) parceiro.setAtivo(request.ativo());
        if (request.verificado() != null) parceiro.setVerificado(request.verificado());

        Parceiro saved = parceiroRepository.save(parceiro);
        return ParceiroResponse.from(saved);
    }

    public ParceiroResponse getById(UUID id) {
        Parceiro parceiro = parceiroRepository.findById(id)
            .orElseThrow(() -> new EntityNotFoundException("Parceiro nao encontrado: " + id));
        return ParceiroResponse.from(parceiro);
    }

    public Page<ParceiroListResponse> list(
        String search,
        CategoriaParceiro categoria,
        String cidade,
        String estado,
        Boolean ativo,
        Boolean verificado,
        Pageable pageable
    ) {
        String categoriaStr = categoria != null ? categoria.name() : null;
        return parceiroRepository.findWithFilters(search, categoriaStr, cidade, estado, ativo, verificado, pageable)
            .map(ParceiroListResponse::from);
    }

    @Transactional
    public void delete(UUID id) {
        Parceiro parceiro = parceiroRepository.findById(id)
            .orElseThrow(() -> new EntityNotFoundException("Parceiro nao encontrado: " + id));
        parceiroRepository.delete(parceiro);
    }

    @Transactional
    public void deactivate(UUID id) {
        Parceiro parceiro = parceiroRepository.findById(id)
            .orElseThrow(() -> new EntityNotFoundException("Parceiro nao encontrado: " + id));
        parceiro.setAtivo(false);
        parceiroRepository.save(parceiro);
    }

    @Transactional
    public void activate(UUID id) {
        Parceiro parceiro = parceiroRepository.findById(id)
            .orElseThrow(() -> new EntityNotFoundException("Parceiro nao encontrado: " + id));
        parceiro.setAtivo(true);
        parceiroRepository.save(parceiro);
    }

    @Transactional
    public void verify(UUID id) {
        Parceiro parceiro = parceiroRepository.findById(id)
            .orElseThrow(() -> new EntityNotFoundException("Parceiro nao encontrado: " + id));
        parceiro.setVerificado(true);
        parceiroRepository.save(parceiro);
    }

    public List<ParceiroListResponse> findByCategoria(CategoriaParceiro categoria) {
        return parceiroRepository.findByCategoria(categoria).stream()
            .map(ParceiroListResponse::from)
            .toList();
    }

    public List<ParceiroListResponse> findByCategoriaAndCidade(CategoriaParceiro categoria, String cidade) {
        return parceiroRepository.findByCategoriaAndCidade(categoria, cidade).stream()
            .map(ParceiroListResponse::from)
            .toList();
    }

    public List<ParceiroListResponse> findByCategoriaAndEstado(CategoriaParceiro categoria, String estado) {
        return parceiroRepository.findByCategoriaAndEstado(categoria, estado).stream()
            .map(ParceiroListResponse::from)
            .toList();
    }

    public List<ParceiroListResponse> findTopRated(int limit) {
        return parceiroRepository.findTopRated(PageRequest.of(0, limit)).stream()
            .map(ParceiroListResponse::from)
            .toList();
    }

    public List<ParceiroListResponse> findTopRatedByCategoria(CategoriaParceiro categoria, int limit) {
        return parceiroRepository.findTopRatedByCategoria(categoria, PageRequest.of(0, limit)).stream()
            .map(ParceiroListResponse::from)
            .toList();
    }

    public Map<CategoriaParceiro, Long> countByCategoria() {
        Map<CategoriaParceiro, Long> counts = new EnumMap<>(CategoriaParceiro.class);
        for (CategoriaParceiro categoria : CategoriaParceiro.values()) {
            counts.put(categoria, parceiroRepository.countByCategoria(categoria));
        }
        return counts;
    }

    public List<CategoriaParceiro> getCategorias() {
        return Arrays.asList(CategoriaParceiro.values());
    }

    private void mapRequestToEntity(CreateParceiroRequest request, Parceiro parceiro) {
        parceiro.setNome(request.nome());
        parceiro.setNomeFantasia(request.nomeFantasia());
        parceiro.setCnpj(request.cnpj());
        parceiro.setCpf(request.cpf());
        parceiro.setEmail(request.email());
        parceiro.setTelefone(request.telefone());
        parceiro.setCelular(request.celular());
        parceiro.setWebsite(request.website());
        parceiro.setEndereco(request.endereco());
        parceiro.setNumero(request.numero());
        parceiro.setComplemento(request.complemento());
        parceiro.setBairro(request.bairro());
        parceiro.setCidade(request.cidade());
        parceiro.setEstado(request.estado());
        parceiro.setCep(request.cep());
        parceiro.setCategorias(request.categorias() != null ? request.categorias() : new HashSet<>());
        parceiro.setDescricaoServicos(request.descricaoServicos());
        parceiro.setAreaAtuacao(request.areaAtuacao());
        parceiro.setContatoNome(request.contatoNome());
        parceiro.setContatoCargo(request.contatoCargo());
        parceiro.setObservacoes(request.observacoes());
        parceiro.setLogoUrl(request.logoUrl());
    }
}

File: frontend/src/services/parceiroService.ts
import { api } from '@/lib/api'
import {
  ParceiroResponse,
  ParceiroListResponse,
  CreateParceiroRequest,
  UpdateParceiroRequest,
  ParceiroFilter,
  CategoriaParceiro,
  CategoriaParceiroResponse,
  Page,
} from '@/types'

const BASE_URL = '/parceiros'

export interface PaginationParams {
  page?: number
  size?: number
  sort?: string
}

export const parceiroService = {
  async create(data: CreateParceiroRequest): Promise<ParceiroResponse> {
    const response = await api.post<ParceiroResponse>(BASE_URL, data)
    return response.data
  },

  async getById(id: string): Promise<ParceiroResponse> {
    const response = await api.get<ParceiroResponse>(`${BASE_URL}/${id}`)
    return response.data
  },

  async update(id: string, data: UpdateParceiroRequest): Promise<ParceiroResponse> {
    const response = await api.put<ParceiroResponse>(`${BASE_URL}/${id}`, data)
    return response.data
  },

  async delete(id: string): Promise<void> {
    await api.delete(`${BASE_URL}/${id}`)
  },

  async list(
    filter?: ParceiroFilter,
    pagination?: PaginationParams
  ): Promise<Page<ParceiroListResponse>> {
    const params = new URLSearchParams()

    if (pagination?.page !== undefined) params.append('page', String(pagination.page))
    if (pagination?.size !== undefined) params.append('size', String(pagination.size))
    if (pagination?.sort) params.append('sort', pagination.sort)

    if (filter?.search) params.append('search', filter.search)
    if (filter?.categoria) params.append('categoria', filter.categoria)
    if (filter?.cidade) params.append('cidade', filter.cidade)
    if (filter?.estado) params.append('estado', filter.estado)
    if (filter?.ativo !== undefined) params.append('ativo', String(filter.ativo))
    if (filter?.verificado !== undefined) params.append('verificado', String(filter.verificado))

    const response = await api.get<Page<ParceiroListResponse>>(
      `${BASE_URL}?${params.toString()}`
    )
    return response.data
  },

  async activate(id: string): Promise<void> {
    await api.post(`${BASE_URL}/${id}/ativar`)
  },

  async deactivate(id: string): Promise<void> {
    await api.post(`${BASE_URL}/${id}/desativar`)
  },

  async verify(id: string): Promise<void> {
    await api.post(`${BASE_URL}/${id}/verificar`)
  },

  async findByCategoria(categoria: CategoriaParceiro): Promise<ParceiroListResponse[]> {
    const response = await api.get<ParceiroListResponse[]>(
      `${BASE_URL}/categoria/${categoria}`
    )
    return response.data
  },

  async findByCategoriaAndCidade(
    categoria: CategoriaParceiro,
    cidade: string
  ): Promise<ParceiroListResponse[]> {
    const response = await api.get<ParceiroListResponse[]>(
      `${BASE_URL}/categoria/${categoria}/cidade/${encodeURIComponent(cidade)}`
    )
    return response.data
  },

  async findByCategoriaAndEstado(
    categoria: CategoriaParceiro,
    estado: string
  ): Promise<ParceiroListResponse[]> {
    const response = await api.get<ParceiroListResponse[]>(
      `${BASE_URL}/categoria/${categoria}/estado/${estado}`
    )
    return response.data
  },

  async findTopRated(limit: number = 10): Promise<ParceiroListResponse[]> {
    const response = await api.get<ParceiroListResponse[]>(
      `${BASE_URL}/top-rated?limit=${limit}`
    )
    return response.data
  },

  async findTopRatedByCategoria(
    categoria: CategoriaParceiro,
    limit: number = 10
  ): Promise<ParceiroListResponse[]> {
    const response = await api.get<ParceiroListResponse[]>(
      `${BASE_URL}/top-rated/categoria/${categoria}?limit=${limit}`
    )
    return response.data
  },

  async countByCategoria(): Promise<Record<CategoriaParceiro, number>> {
    const response = await api.get<Record<CategoriaParceiro, number>>(
      `${BASE_URL}/stats/categorias`
    )
    return response.data
  },

  async getCategorias(): Promise<CategoriaParceiroResponse[]> {
    const response = await api.get<CategoriaParceiroResponse[]>(
      `${BASE_URL}/categorias`
    )
    return response.data
  },

  async gerarOfertaIA(parceiroId: string, condominioId: string): Promise<string> {
    const response = await api.post<{ texto: string }>(
      `${BASE_URL}/${parceiroId}/oferta-ia`,
      { condominioId }
    )
    return response.data.texto
  },
}

export default parceiroService
